### DISCUSSION TITLE
document dynamic/async completion

### DISCUSSION THREAD

**Discussion by mooseyboots** [minad/consult#1123] [Ideas]
it'd be nice if there were some documentation of how to dynamically generate candidates for `consult--read`, for package authors using consult.

it's referred to as both async and dynamic, its not clear if there is a difference?

i couldn't find any mention of `consult--dynamic-completion` in the readme or the wiki. and its docstring is terse.

meanwhile the in-built consult async functions seem focused on CLI cmds, which was not my use case.

i had to read the code of an external package that uses consult to work out what to do.

**Comment by minad** [comment]
For now you can take a look at the source of `consult-info`.

Neither the async nor the dynamic API are well documented. This APIs are also private. It may make sense to change this at some point, making the underlying completion APIs public or factoring them out.

**Comment by mooseyboots** [comment]
thanks for your response.

yeah, i figured it is worth documenting somehow and/or making public, as it is also its stated raison d'etre, as per its docstring: "You should use completing-read instead of consult--read if you don't use asynchronous candidate computation or candidate preview."

**Reply by minad** [reply]
> ...as it is also its stated raison d'etre, as per its docstring...

Well, to be clear, the raison d'être is internal use in Consult. But many people want to use the APIs nevertheless. If you want to help writing documentation, you are very welcome. :)

**Reply by vermiculus** [reply]
I recently had cause to create some basic async completion and I've got it in a pretty neat/tidy example, so I can take a first crack at writing some docs. I've submitted #1130.

**Reply by vermiculus** [reply]
@mooseyboots FYI, that PR has now been merged :-)
